massa-final-state/src/executed_ops.rs

```diff
@@ -22,39 +25,66 @@ use std::ops::Bound::{Excluded, Included};
 
 /// A structure to list and prune previously executed operations
 #[derive(Debug, Default, Clone, PartialEq, Eq)]
-pub struct ExecutedOps(PreHashMap<OperationId, Slot>);
+pub struct ExecutedOps {
+    /// Map of the executed operations
+    ops: PreHashMap<OperationId, Slot>,
+    /// Cumulated hash of the executed operations
+    pub hash: Option<Hash>,
+}
 
 impl ExecutedOps {
     /// returns the number of executed operations
     pub fn len(&self) -> usize {
-        self.0.len()
+        self.ops.len()
     }
 
     /// Check is there is no executed ops
     pub fn is_empty(&self) -> bool {
-        self.0.is_empty()
+        self.ops.is_empty()
     }
 
     /// extends with another ExecutedOps
     pub fn extend(&mut self, other: ExecutedOps) {
-        self.0.extend(other.0);
+        if let Some(other_hash) = other.hash {
+            if let Some(current_hash) = self.hash.as_mut() {
+                *current_hash ^= other_hash;
+            } else {
+                self.hash = Some(other_hash);
+            }
+        }
+        self.ops.extend(other.ops);
     }
 
     /// check if an operation was executed
     pub fn contains(&self, op_id: &OperationId) -> bool {
-        self.0.contains_key(op_id)
+        self.ops.contains_key(op_id)
     }
 
     /// marks an op as executed
     pub fn insert(&mut self, op_id: OperationId, last_valid_slot: Slot) {
-        self.0.insert(op_id, last_valid_slot);
+        if let Some(current_hash) = self.hash.as_mut() {
+            *current_hash ^= *op_id.get_hash();
+        } else {
+            self.hash = Some(*op_id.get_hash());
+        }
+        self.ops.insert(op_id, last_valid_slot);
     }
 
     /// Prune all operations that expire strictly before max_slot
     pub fn prune(&mut self, max_slot: Slot) {
         // TODO use slot-sorted structure for more efficient pruning (this has a linear complexity currently)
-        self.0
-            .retain(|_id, last_valid_slot| *last_valid_slot >= max_slot);
+        let (kept, removed): (PreHashMap<OperationId, Slot>, PreHashMap<OperationId, Slot>) = self
+            .ops
+            .iter()
+            .partition(|(_, &last_valid_slot)| last_valid_slot >= max_slot);
+        let hash = self
+            .hash
+            .as_mut()
+            .expect("critical: an ExecutedOps object with ops must also contain a hash");
```

Comment:
This will fail if prune is called and no operations as been inserted yet ?

Reply:
Added the following check right above:
```rust
if removed.is_empty() {
    return;
}
```